### ISSUE TITLE
Q on write amplification

### ISSUE THREAD

**Issue by glycerine** [couchbase/moss#44]
@mschoch I enjoyed your talk on Moss at GopherCon.  At the end you pointed out a situation (I couldn't discern exact when) where a small write lead to alot of copying/write-amplification.

I just wanted to inquire if that issue had been addressed?

**Comment by mschoch** [comment]
Thanks,

Yes, the issue I mentioned in the slides was related to when we decide to perform compaction.  I didn't explain this in the talk, but it was configurable by a value.  The calculation of the value isn't so important, but when the value was 0, it meant we compact whenever we can.  Setting it to non-zero meant it would sometimes choose not to compact, but in practice, setting it to a good value was hard to do.  Instead, we observed better behavior when it was 0.

However, the value of 0 would lead to overly aggressive compaction, even when the situation didn't really warrant it.

The solution was to change the compaction behavior, and also the metric by which we choose/tune the compaction process.

Many of the recent commits by @hisundar have been addressing this problem, by moving us towards a leveled compaction model.   At a very high levels means preferring to leave larger segments as is, and preferring to combine smaller segments.  This has two big benefits, the newly merged segments are smaller (when possible) and the larger segments can stay in the cache longer.

So, I would say the specific problem I shared has been solved.  And in general the write-amplication issues are much better now than they were, but as always there is more work to be done.

If you have more questions I'm sure @hisundar can help.
